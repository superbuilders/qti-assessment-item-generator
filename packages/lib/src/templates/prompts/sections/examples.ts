import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
