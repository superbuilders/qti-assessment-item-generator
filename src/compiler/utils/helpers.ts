import { typedSchemas } from "@/widgets/registry"

/**
 * Calculates the greatest common divisor (GCD) of two numbers using the Euclidean algorithm.
 * @param a The first number.
 * @param b The second number.
 * @returns The GCD of a and b.
 */
export function gcd(a: number, b: number): number {
	return b === 0 ? a : gcd(b, a % b)
}

/**
 * Determines if a fraction results in a terminating decimal.
 * A fraction terminates if and only if the prime factors of its simplified denominator are only 2s and 5s.
 * @param numerator The numerator of the fraction.
 * @param denominator The denominator of the fraction.
 * @returns True if the fraction terminates, false otherwise.
 */
export function isTerminatingFraction(
	numerator: number,
	denominator: number
): boolean {
	if (denominator === 0) {
		return false
	}
	// Simplify the fraction by dividing by the GCD.
	const commonDivisor = gcd(Math.abs(numerator), Math.abs(denominator))
	let simplifiedDen = Math.abs(denominator) / commonDivisor

	// The fraction terminates if the simplified denominator's prime factors are only 2 and 5.
	while (simplifiedDen % 2 === 0) {
		simplifiedDen /= 2
	}
	while (simplifiedDen % 5 === 0) {
		simplifiedDen /= 5
	}

	return simplifiedDen === 1
}

export function encodeDataUri(content: string): string {
	const base64 = btoa(unescape(encodeURIComponent(content)))
	const isSvg = content.trim().startsWith("<svg")
	return `${isSvg ? "data:image/svg+xml" : "data:text/html"};base64,${base64}`
}

// Helper function to check if a string is a valid widget type
export function isValidWidgetType(
	type: string
): type is keyof typeof typedSchemas {
	return type in typedSchemas
}

/**
 * Attempts to approximate simple multiplicative/divisive expressions that include π or pi
 * by replacing pi with the provided approximation (default 3.14) and evaluating the expression.
 *
 * Supported forms (case-insensitive for "pi"):
 * - "144*pi", "pi*144", "144pi" (implicit multiplication), "pi144" (rare, but treated as pi*144)
 * - Chains with '*' and '/' only, e.g. "3/2*pi", "pi/4", "6*pi/5", "(144) * (pi)"
 * Parentheses are allowed but only for grouping; no addition/subtraction is supported.
 *
 * Returns a normalized decimal string if evaluation succeeds, otherwise null.
 */
export function tryApproximatePiProduct(
	rawExpr: string,
	piApprox = 3.14
): string | null {
	if (typeof rawExpr !== "string" || rawExpr.trim() === "") return null

	// Quick check: must contain a pi token of some sort
	const containsPi = /\bpi\b/i.test(rawExpr) || rawExpr.includes("π")
	if (!containsPi) return null

	// Normalize: insert explicit * for implicit numeric-pi adjacency in common cases
	const expr = rawExpr
		// insert * between number and pi (e.g., 144pi -> 144*pi)
		.replace(/(\d)\s*(π|\bpi\b)/gi, "$1*$2")
		// and between pi and number (e.g., pi144 -> pi*144)
		.replace(/(π|\bpi\b)\s*(\d)/gi, "$1*$2")
		// remove whitespace
		.replace(/\s+/g, "")
		// drop parentheses (multiplicative-only evaluation does not require them)
		.replace(/[()]/g, "")
		// normalize 'pi' variants to numeric approximation
		.replace(/π/gi, String(piApprox))
		.replace(/\bpi\b/gi, String(piApprox))

	// After substitution, we only allow digits, '.', '*', '/' characters
	if (/[^0-9.*/]/.test(expr)) return null
	// Disallow addition/subtraction in this lightweight evaluator
	if (/[+-]/.test(expr)) return null
	// Must still contain the approximation value to ensure we actually used pi
	if (!expr.includes(String(piApprox))) return null

	// Split into tokens by * and / and evaluate left-to-right (* and / are same precedence)
	const parts: string[] = expr.split(/([*/])/)
	if (parts.length === 0) return null

	// Validate tokens
	for (let i = 0; i < parts.length; i++) {
		const t = parts[i]
		if (i % 2 === 0) {
			// operand
			if (t === "" || Number.isNaN(Number(t))) return null
		} else {
			// operator must be * or /
			if (t !== "*" && t !== "/") return null
		}
	}

	let acc = Number(parts[0])
	for (let i = 1; i < parts.length; i += 2) {
		const op = parts[i]
		const rhs = Number(parts[i + 1])
		if (op === "*") acc = acc * rhs
		else {
			if (rhs === 0) return null
			acc = acc / rhs
		}
	}

	// Normalize number to a compact string while preserving significant digits
	// Avoid scientific notation for typical classroom values
	const normalizeNumberString = () => {
		const maybeFixed = acc.toFixed(10) // limit noise
		// trim trailing zeros
		const trimmed = maybeFixed.replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1")
		return trimmed
	}
	const asString = normalizeNumberString()

	// Add a guard: must be a finite number
	if (!Number.isFinite(Number(asString))) return null
	return asString
}

/**
 * Normalizes a string part for use in a derived QTI identifier.
 * Converts to upper snake case and removes invalid characters.
 * @param part The string part to normalize.
 * @returns The normalized string.
 */
export function normalizeIdPart(part: string): string {
	return part.toUpperCase().replace(/[^A-Z0-9_]/g, "_")
}

/**
 * Derives a deterministic, QTI-safe combo identifier from a path of response parts.
 * @param pathParts An array of strings, where each string is a response-value pair (e.g., "RESPONSE_1_A").
 * @returns The fully-formed combo identifier (e.g., "FB__RESPONSE_1_A__RESPONSE_2_CORRECT").
 */
export function deriveComboIdentifier(pathParts: string[]): string {
	return `FB__${pathParts.join("__")}`
}
