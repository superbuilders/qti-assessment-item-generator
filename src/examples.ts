import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
