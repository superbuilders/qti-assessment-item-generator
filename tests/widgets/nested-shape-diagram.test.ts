import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { nestedShapeDiagramExamples } from "../../examples/nested-shape-diagram"
import { generateWidgetForTest } from "../helpers/generateWidgetForTest"

describe("nestedShapeDiagram widget tests", () => {
	// Test each example from the examples file
	nestedShapeDiagramExamples.forEach((example, index) => {
		// Extract question ID from the example if available (from comments in the source)
		const questionIds = [
			"xdd3424e759f98328",
			"x9b74005f28d0b824",
			"x70ade9874035cc08",
			"x7893c043131fd6d6",
			"xd8c0fb6d8bfff96e",
			"x3a68667079fdffdd",
			"x55061cb6ed4c4ee5",
			"xb9a2f2df68f7f278",
			"xc12b923ac5d6ad9e",
			"x9b473530bde85c53",
			"xf0438c2e9eb979cd",
			"x74eb96d7a07de3a4"
		]

		const descriptions = [
			"Area of a shaded region between a circle and a rectangle",
			"Area of a shaded region (circle minus rectangle)",
			"Area of a shaded region (square minus circle)",
			"Area of a shaded region between a circle and a rectangle",
			"Area of a shaded region (circle minus rectangle)",
			"Area of the shaded region (circle minus rectangle)",
			"Area of the shaded region between a rectangle and a circle",
			"Area of the shaded region: rectangle minus circle",
			"Area of a shaded region (rectangle with inscribed circle)",
			"Area of a shaded region: rectangle minus circle",
			"Area of the shaded region between a square and a circle",
			"Area of a shaded region between a circle and a square"
		]

		const testName = questionIds[index]
			? `nested-shape-diagram - [${questionIds[index]}] ${descriptions[index]}`
			: `nested-shape-diagram - Example ${index + 1}`

		test(testName, async () => {
			const result = await errors.try(generateWidgetForTest(example))
			if (result.error) {
				logger.error("widget generation failed for nestedShapeDiagram", {
					error: result.error,
					exampleIndex: index,
					inputData: example
				})
				throw errors.wrap(result.error, "widget generation")
			}

			const svg = result.data
			expect(svg).toMatchSnapshot()
		})
	})
})
