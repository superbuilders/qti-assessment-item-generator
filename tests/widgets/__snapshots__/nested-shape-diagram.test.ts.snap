// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [xdd3424e759f98328] Area of a shaded region between a circle and a rectangle 1`] = `"<svg width="240" height="240" viewBox="-131 -131 391 391" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 110 0 A 110 110 0 1 1 -110 0 A 110 110 0 1 1 110 0 Z M -20 -30 L 20 -30 L 20 30 L -20 30 Z" fill="#ccfaff80" fill-rule="evenodd"/><circle cx="0" cy="0" r="110" fill="none" stroke="#333333" stroke-width="1.5"/><polyline points="-20,-30 20,-30 20,30 -20,30" stroke="#333333" stroke-width="1.5" fill="none"/><line x1="0" y1="0" x2="110" y2="0" stroke="#ff0000" stroke-width="1.5"/><text x="61.5" y="-10" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">11 cm</text><line x1="-20" y1="-30" x2="20" y2="-30" stroke="#333333" stroke-width="1.5"/><text x="0" y="-40" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">6 cm</text><line x1="20" y1="-30" x2="20" y2="30" stroke="#333333" stroke-width="1.5"/><line x1="20" y1="30" x2="-20" y2="30" stroke="#333333" stroke-width="1.5"/><line x1="-20" y1="30" x2="-20" y2="-30" stroke="#333333" stroke-width="1.5"/><text x="-35" y="0" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">4 cm</text></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [x9b74005f28d0b824] Area of a shaded region (circle minus rectangle) 1`] = `"<svg width="260" height="260" viewBox="-141 -141 421 421" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 120 0 A 120 120 0 1 1 -120 0 A 120 120 0 1 1 120 0 Z M -25 -55 L 25 -55 L 25 55 L -25 55 Z" fill="#ccfaff80" fill-rule="evenodd"/><circle cx="0" cy="0" r="120" fill="none" stroke="#333333" stroke-width="1.5"/><polyline points="-25,-55 25,-55 25,55 -25,55" stroke="#333333" stroke-width="1.5" fill="none"/><line x1="0" y1="0" x2="120" y2="0" stroke="#ff0000" stroke-width="1.5"/><text x="69" y="-10" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">12 cm</text><line x1="-25" y1="-55" x2="25" y2="-55" stroke="#333333" stroke-width="1.5"/><text x="0" y="-65" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">5 cm</text><line x1="25" y1="-55" x2="25" y2="55" stroke="#333333" stroke-width="1.5"/><line x1="25" y1="55" x2="-25" y2="55" stroke="#333333" stroke-width="1.5"/><line x1="-25" y1="55" x2="-25" y2="-55" stroke="#333333" stroke-width="1.5"/><text x="-40" y="0" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">11 cm</text></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [x70ade9874035cc08] Area of a shaded region (square minus circle) 1`] = `"<svg width="210" height="175" viewBox="-20 -20 250 215" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 35 35 L 140 35 L 140 140 L 35 140 Z M 122.5 87.5 A 35 35 0 1 1 52.5 87.5 A 35 35 0 1 1 122.5 87.5 Z" fill="#ccfaff80" fill-rule="evenodd"/><polyline points="35,35 140,35 140,140 35,140" stroke="#333333" stroke-width="1.5" fill="none"/><circle cx="87.5" cy="87.5" r="35" fill="none" stroke="#333333" stroke-width="1.5"/><line x1="87.5" y1="87.5" x2="122.5" y2="87.5" stroke="#ff0000" stroke-width="1.5"/><text x="105" y="77.5" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">1 cm</text><line x1="35" y1="35" x2="140" y2="35" stroke="#333333" stroke-width="1.5"/><text x="87.5" y="25" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">3 cm</text><line x1="140" y1="35" x2="140" y2="140" stroke="#333333" stroke-width="1.5"/><text x="155" y="87.5" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">3 cm</text><line x1="140" y1="140" x2="35" y2="140" stroke="#333333" stroke-width="1.5"/><line x1="35" y1="140" x2="35" y2="35" stroke="#333333" stroke-width="1.5"/></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [x7893c043131fd6d6] Area of a shaded region between a circle and a rectangle 1`] = `"<svg width="360" height="360" viewBox="-20 -20 400 400" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 340 180 A 160 160 0 1 1 20 180 A 160 160 0 1 1 340 180 Z M 60 160 L 300 160 L 300 200 L 60 200 Z" fill="#ccfaff80" fill-rule="evenodd"/><circle cx="180" cy="180" r="160" fill="none" stroke="#333333" stroke-width="1.5"/><polyline points="60,160 300,160 300,200 60,200" stroke="#333333" stroke-width="1.5" fill="none"/><line x1="180" y1="180" x2="340" y2="180" stroke="#ff0000" stroke-width="1.5"/><text x="242.5" y="170" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">8 cm</text><line x1="60" y1="160" x2="300" y2="160" stroke="#333333" stroke-width="1.5"/><text x="180" y="150" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">12 cm</text><line x1="300" y1="160" x2="300" y2="200" stroke="#333333" stroke-width="1.5"/><line x1="300" y1="200" x2="60" y2="200" stroke="#333333" stroke-width="1.5"/><line x1="60" y1="200" x2="60" y2="160" stroke="#333333" stroke-width="1.5"/><text x="45" y="180" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">2 cm</text></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [xd8c0fb6d8bfff96e] Area of a shaded region (circle minus rectangle) 1`] = `"<svg width="280" height="280" viewBox="-20 -20 320 320" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 260 140 A 120 120 0 1 1 20 140 A 120 120 0 1 1 260 140 Z M 70 90 L 210 90 L 210 190 L 70 190 Z" fill="#ccfaff" fill-rule="evenodd"/><circle cx="140" cy="140" r="120" fill="none" stroke="#333333" stroke-width="1.5"/><polyline points="70,90 210,90 210,190 70,190" stroke="#333333" stroke-width="1.5" fill="none"/><line x1="140" y1="140" x2="260" y2="140" stroke="#ff0000" stroke-width="1.5"/><text x="177.5" y="130" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">6 cm</text><line x1="70" y1="90" x2="210" y2="90" stroke="#333333" stroke-width="1.5"/><text x="140" y="80" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">7 cm</text><line x1="210" y1="90" x2="210" y2="190" stroke="#333333" stroke-width="1.5"/><line x1="210" y1="190" x2="70" y2="190" stroke="#333333" stroke-width="1.5"/><line x1="70" y1="190" x2="70" y2="90" stroke="#333333" stroke-width="1.5"/><text x="55" y="140" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">5 cm</text></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [x3a68667079fdffdd] Area of the shaded region (circle minus rectangle) 1`] = `"<svg width="200" height="200" viewBox="-20 -20 240 240" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 180 100 A 80 80 0 1 1 20 100 A 80 80 0 1 1 180 100 Z M 70 80 L 130 80 L 130 120 L 70 120 Z" fill="#ccfaff80" fill-rule="evenodd"/><circle cx="100" cy="100" r="80" fill="none" stroke="#333333" stroke-width="1.5"/><polyline points="70,80 130,80 130,120 70,120" stroke="#333333" stroke-width="1.5" fill="none"/><line x1="100" y1="100" x2="180" y2="100" stroke="#ff0000" stroke-width="1.5"/><text x="151.5" y="90" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">4 cm</text><line x1="70" y1="80" x2="130" y2="80" stroke="#333333" stroke-width="1.5"/><text x="100" y="70" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">3 cm</text><line x1="130" y1="80" x2="130" y2="120" stroke="#333333" stroke-width="1.5"/><line x1="130" y1="120" x2="70" y2="120" stroke="#333333" stroke-width="1.5"/><line x1="70" y1="120" x2="70" y2="80" stroke="#333333" stroke-width="1.5"/><text x="55" y="100" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">2 cm</text></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [x55061cb6ed4c4ee5] Area of the shaded region between a rectangle and a circle 1`] = `"<svg width="350" height="350" viewBox="-20 -20 390 390" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 25 25 L 300 25 L 300 325 L 25 325 Z M 187.5 175 A 25 25 0 1 1 137.5 175 A 25 25 0 1 1 187.5 175 Z" fill="#ccfaff80" fill-rule="evenodd"/><polyline points="25,25 300,25 300,325 25,325" stroke="#333333" stroke-width="1.5" fill="none"/><circle cx="162.5" cy="175" r="25" fill="none" stroke="#333333" stroke-width="1.5"/><line x1="162.5" y1="175" x2="187.5" y2="175" stroke="#ff0000" stroke-width="1.5"/><text x="175" y="165" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">1 cm</text><line x1="25" y1="25" x2="300" y2="25" stroke="#333333" stroke-width="1.5"/><text x="162.5" y="15" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">11 cm</text><line x1="300" y1="25" x2="300" y2="325" stroke="#333333" stroke-width="1.5"/><text x="315" y="175" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">12 cm</text><line x1="300" y1="325" x2="25" y2="325" stroke="#333333" stroke-width="1.5"/><line x1="25" y1="325" x2="25" y2="25" stroke="#333333" stroke-width="1.5"/></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [xb9a2f2df68f7f278] Area of the shaded region: rectangle minus circle 1`] = `"<svg width="275" height="225" viewBox="-20 -20 315 265" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 25 25 L 225 25 L 225 200 L 25 200 Z M 200 112.5 A 75 75 0 1 1 50 112.5 A 75 75 0 1 1 200 112.5 Z" fill="#ccfaff80" fill-rule="evenodd"/><polyline points="25,25 225,25 225,200 25,200" stroke="#333333" stroke-width="1.5" fill="none"/><circle cx="125" cy="112.5" r="75" fill="none" stroke="#333333" stroke-width="1.5"/><line x1="125" y1="112.5" x2="200" y2="112.5" stroke="#ff0000" stroke-width="1.5"/><text x="161.5" y="102.5" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">3 cm</text><line x1="25" y1="25" x2="225" y2="25" stroke="#333333" stroke-width="1.5"/><text x="125" y="15" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">8 cm</text><line x1="225" y1="25" x2="225" y2="200" stroke="#333333" stroke-width="1.5"/><text x="240" y="112.5" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">7 cm</text><line x1="225" y1="200" x2="25" y2="200" stroke="#333333" stroke-width="1.5"/><line x1="25" y1="200" x2="25" y2="25" stroke="#333333" stroke-width="1.5"/></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [xc12b923ac5d6ad9e] Area of a shaded region (rectangle with inscribed circle) 1`] = `"<svg width="210" height="210" viewBox="-20 -20 250 250" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 35 35 L 140 35 L 140 175 L 35 175 Z M 122.5 105 A 35 35 0 1 1 52.5 105 A 35 35 0 1 1 122.5 105 Z" fill="#ccfaff80" fill-rule="evenodd"/><polyline points="35,35 140,35 140,175 35,175" stroke="#333333" stroke-width="1.5" fill="none"/><circle cx="87.5" cy="105" r="35" fill="none" stroke="#333333" stroke-width="1.5"/><line x1="87.5" y1="105" x2="122.5" y2="105" stroke="#ff0000" stroke-width="1.5"/><text x="105" y="95" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">1 cm</text><line x1="35" y1="35" x2="140" y2="35" stroke="#333333" stroke-width="1.5"/><text x="87.5" y="25" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">3 cm</text><line x1="140" y1="35" x2="140" y2="175" stroke="#333333" stroke-width="1.5"/><text x="155" y="105" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">4 cm</text><line x1="140" y1="175" x2="35" y2="175" stroke="#333333" stroke-width="1.5"/><line x1="35" y1="175" x2="35" y2="35" stroke="#333333" stroke-width="1.5"/></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [x9b473530bde85c53] Area of a shaded region: rectangle minus circle 1`] = `"<svg width="350" height="350" viewBox="-20 -20 390 390" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 25 25 L 300 25 L 300 325 L 25 325 Z M 212.5 175 A 50 50 0 1 1 112.5 175 A 50 50 0 1 1 212.5 175 Z" fill="#ccfaff80" fill-rule="evenodd"/><polyline points="25,25 300,25 300,325 25,325" stroke="#333333" stroke-width="1.5" fill="none"/><circle cx="162.5" cy="175" r="50" fill="none" stroke="#333333" stroke-width="1.5"/><line x1="162.5" y1="175" x2="212.5" y2="175" stroke="#ff0000" stroke-width="1.5"/><text x="186.5" y="165" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">2 cm</text><line x1="25" y1="25" x2="300" y2="25" stroke="#333333" stroke-width="1.5"/><text x="162.5" y="15" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">11 cm</text><line x1="300" y1="25" x2="300" y2="325" stroke="#333333" stroke-width="1.5"/><text x="315" y="175" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">12 cm</text><line x1="300" y1="325" x2="25" y2="325" stroke="#333333" stroke-width="1.5"/><line x1="25" y1="325" x2="25" y2="25" stroke="#333333" stroke-width="1.5"/></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [xf0438c2e9eb979cd] Area of the shaded region between a square and a circle 1`] = `"<svg width="350" height="325" viewBox="-20 -20 390 365" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 25 25 L 300 25 L 300 300 L 25 300 Z M 287.5 162.5 A 125 125 0 1 1 37.5 162.5 A 125 125 0 1 1 287.5 162.5 Z" fill="#ccfaff80" fill-rule="evenodd"/><polyline points="25,25 300,25 300,300 25,300" stroke="#333333" stroke-width="1.5" fill="none"/><circle cx="162.5" cy="162.5" r="125" fill="none" stroke="#333333" stroke-width="1.5"/><line x1="162.5" y1="162.5" x2="287.5" y2="162.5" stroke="#ff0000" stroke-width="1.5"/><text x="224" y="152.5" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">5 cm</text><line x1="25" y1="25" x2="300" y2="25" stroke="#333333" stroke-width="1.5"/><text x="162.5" y="15" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">11 cm</text><line x1="300" y1="25" x2="300" y2="300" stroke="#333333" stroke-width="1.5"/><text x="315" y="162.5" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">11 cm</text><line x1="300" y1="300" x2="25" y2="300" stroke="#333333" stroke-width="1.5"/><line x1="25" y1="300" x2="25" y2="25" stroke="#333333" stroke-width="1.5"/></svg>"`;

exports[`nestedShapeDiagram widget tests nested-shape-diagram - [x74eb96d7a07de3a4] Area of a shaded region between a circle and a square 1`] = `"<svg width="200" height="200" viewBox="-20 -20 240 240" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="11"><path d="M 180 100 A 80 80 0 1 1 20 100 A 80 80 0 1 1 180 100 Z M 70 70 L 130 70 L 130 130 L 70 130 Z" fill="#ccfaff" fill-rule="evenodd"/><circle cx="100" cy="100" r="80" fill="none" stroke="#333333" stroke-width="1.5"/><polyline points="70,70 130,70 130,130 70,130" stroke="#333333" stroke-width="1.5" fill="none"/><line x1="70" y1="70" x2="130" y2="70" stroke="#333333" stroke-width="1.5"/><text x="100" y="60" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">3 cm</text><line x1="130" y1="70" x2="130" y2="130" stroke="#333333" stroke-width="1.5"/><line x1="130" y1="130" x2="70" y2="130" stroke="#333333" stroke-width="1.5"/><line x1="70" y1="130" x2="70" y2="70" stroke="#333333" stroke-width="1.5"/><text x="55" y="100" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#333333" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">3 cm</text><line x1="100" y1="100" x2="180" y2="100" stroke="#ff0000" stroke-width="1.5"/><text x="151.5" y="90" font-size="11" text-anchor="middle" dominant-baseline="middle" fill="#ff0000" stroke="#fafafa" stroke-width="0.2em" paint-order="stroke fill">4 cm</text></svg>"`;
